### ISSUE TITLE
cofit_arrow demands bit-idempotence from a warm-started iterative solver — and the fall-through fix for the silent Cholesky no-op (reverted in e55ed1d33) is blocked on it

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#2508]
## Summary

Two contracts are propped up by a single bug, in opposite directions, so fixing it flips one test green and another red. The fix for the silent no-op is written and verified; it is blocked on a decision about `cofit_arrow`'s idempotence predicate, which belongs to that contract's owner rather than to the solver.

History: landed as `587fa42c7`, reverted as `e55ed1d33` when the regression surfaced. Main is clean.

## The bug being fixed

`crates/gam-sae/src/sparse_dict/update.rs`, small-component decoder solve:

```rust
let Some(sol) = cholesky_solve_block(&mat, &rhs) else {
    stats.dense_factorization_failures += 1;
    return Ok(());          // no solution written, Ok returned
};
```

On a declined Cholesky this returns **success without solving**, leaving the decoder block at its previous value and skipping the matrix-free block-CG path that sits immediately below and solves the same SPD system. The only trace is a counter.

Downstream, the certification gate reads that counter as a veto (`dense_factorization_failures == 0` inside `numerically_sound`), so a fit is refused as non-convergent with:

```
InnerNonConvergence { epochs: 5, explained_variance: 1.0, ev_residual: 0.0, tolerance: 1e-6,
                      accepted_births: 0, decoder_fixed_point_residual: 0.0, routing_residual: 0.0,
                      solve_residual: 0.0, solve_tolerance: 1.49e-8,
                      decoder_nonconverged_columns: 0, decoder_factorization_failures: 8 }
```

Explained variance exactly 1.0, every residual exactly 0.0 — refused because a faster path declined on the way to the same exact answer. That is `gam-sae::sparse_lane_no_dense_assignment large_k_sparse_fit_stays_fixed_width_and_never_materializes_dense_n_by_k`, and the fix makes it pass (guard-verified, post-run md5 matching the diff).

## Why fixing it breaks idempotence

`sparse_dict::cofit_arrow::tests::arrow_routed_linear_tier_matches_or_beats_block_reconstruction_2023` then fails:

```
cofit_linear_via_arrow: deterministic joint-solver re-entry did not reach an idempotent fixed point
within 8 re-entries of 128 inner iterations each; the last pass was blocked by:
the Newton walk reported it moved model state
```

The mechanism is one line — `update.rs:2627`:

```rust
initial_row[j] = decoder[[a, c]] as f64;
```

The block CG **warm-starts from the current decoder**. Therefore:

- A **direct Cholesky** is idempotent *by construction*: it computes the exact solution of a fixed system regardless of the starting value, so successive passes are bit-identical and re-entry cannot move state.
- A **warm-started iterative solve** is idempotent only *to its termination tolerance*: pass 2 starts from pass 1's output, already near the solution, and stops when its residual test is met — having moved by up to that tolerance.
- **Before the fix**, the affected blocks took neither path. Nothing happened, so nothing moved: idempotence held *vacuously*.

So `arrow_routed_linear_tier...` was passing **because the computation was missing**. One bug, two contracts, opposite directions.

**Selection effect that makes it worse:** the redirected blocks are *by definition* the small components whose Cholesky failed — the ill-conditioned ones. CG on an ill-conditioned system terminates with the loosest realized residual, so their per-re-entry movement is the largest in the whole solve. Blocks above `direct_solve_size_threshold(k)` already used this same warm-started CG and idempotence held for them; what the fix adds is precisely the tail where it does not.

## Options

**(b) "warm-start the CG so re-entry reproduces the same solution" — DEAD.** It already warm-starts; warm-starting is the cause, not the cure. Recorded so nobody re-proposes it.

**(c) Make the idempotence predicate tolerance-aware** — measure "moved model state" against the iterative solver's own realized residual rather than against exactness. **Recommended.** Demanding bit-idempotence from a warm-started iterative solver asks for a guarantee it does not offer, and the predicate only ever held because the ill-conditioned blocks were not being solved at all. This is the fixed-tolerance-vs-achievable-resolution genus: a tolerance must be derived from the resolution of the instrument that measures it.

**(c') Solve the redirected blocks tighter** — drive their CG below the idempotence threshold. Correctness-equivalent, more expensive, and it puts the burden on the solver rather than on a contract that is arguably over-specified.

**Honest note:** (c) *changes `cofit_arrow`'s contract*. That is why this is filed rather than pushed — the solver side should not unilaterally loosen another module's idempotence guarantee to make its own patch fit.

## Reland shape

Fall-through **and** the tolerance-aware predicate land **together**, with both tests green in one receipt:
- `large_k_sparse_fit_stays_fixed_width_and_never_materializes_dense_n_by_k` (currently red)
- `arrow_routed_linear_tier_matches_or_beats_block_reconstruction_2023` (currently green, must stay green)

Deciding measurement for whoever relands, if (c) vs (c') needs adjudicating empirically: instrument the re-entry to print the per-pass decoder delta on CG-solved blocks. Delta at CG tolerance ⇒ (c). Delta shrinking geometrically ⇒ re-entry budget. Delta not shrinking ⇒ a genuine re-entrancy defect in the fall-through.

**Comment by SauersML** [comment]
Causality correction from an exact history/path audit before relanding this:

- `587fa42c7` changed exactly one file: `crates/gam-sae/src/sparse_dict/update.rs`.
- The cited red test executes `cofit_arrow.rs -> fit_block_sparse_dictionary` in `block.rs -> cofit_linear_via_arrow` / the manifold joint solver. `block.rs` has no runtime call into `update.rs`.
- The executable-path blobs are byte-identical in `587fa42c7^` and `587fa42c7`: `block.rs` = `ac60e45cd491142d45ed6994e8e6b5e0b4e0fc61`, `cofit_arrow.rs` = `d2f7fc4f1192a3be101023007887674c1a2c27cf`, `fit_drivers.rs` = `5a3d4d45818e86fef2416baf7c2beb4e8aec1b6e`.

Therefore the cofit failure cannot be a semantic regression caused by the Cholesky fall-through commit. It was pre-existing, nondeterministic, or shared-state evidence. In particular, option (c) is not valid: no sparse-CG telemetry reaches the cofit/manifold path, and residual alone is not a decoder-state error bound on an ill-conditioned block (`e = A^-1 r` requires an inverse-norm/condition certificate).

The solver bug should be relanded independently: a small-block Cholesky decline is route telemetry, then the already-general matrix-free PCG must solve the block; acceptance is governed by the PCG convergence/residual certificate. I am implementing that pair with a non-vacuous large-K regression asserting decline > 0, CG columns > 0, zero nonconverged columns, and realized residual <= the stop.

The separate cofit convergence defect has a different root cause: whole-pass entry-byte recurrence is not a convergence certificate because algorithmic warm state such as `inner_line_search_warm_step` is live but omitted from `SaeManifoldMutableState`. Its principled contract is terminal `NoStrictDecrease` plus exact preservation through post-termination restore/sweep operations—not a tolerance borrowed from this unrelated sparse solver.

**Comment by SauersML** [comment]
🔒 claim: lane streaming-parity (local fleet). Measuring the revert's premise now.

**Comment by SauersML** [comment]
## The revert's premise is refuted: 587fa42c7 is not on the failing test's path

**Static.** `arrow_routed_linear_tier_matches_or_beats_block_reconstruction_2023` runs `fit_block_sparse_dictionary` (`sparse_dict/block.rs`) and then `cofit_linear_via_arrow` (`sparse_dict/cofit_arrow.rs` → manifold). Both hunks of 587fa42c7 are in `sparse_dict/update.rs`, and **block.rs imports nothing from it** — its entire `use` list is

```rust
use super::scoring::TopSSelector;
use crate::frames::GrassmannFrame;
use ndarray::{...}; use rayon::prelude::*; use std::fmt;
```

`solve_component` (the fall-through site) is reachable only through `solve_decoder_recycled` ← `solve_decoder_with_routability_gate_recycled` ← `update::run` (`update.rs:677`) / `stream.rs:392`. The `numerically_sound` clause lives in `update::run` for the same reason. Neither is entered by that test. So the idempotence contract was never reading the changed code.

**Empirical.** Lane `~/lane-strparity` = `989ed8144` + `git revert --no-commit e55ed1d33` (i.e. 587fa42c7 re-applied; the resulting `update.rs` blob hashes `357805d0d2039dfc9cea87a0c506f2deb3620d10`, which is also exactly what I would land — reconstructed independently on the Mac from `git show $BASE:…` plus the patch, same hash). A10, `RAYON_NUM_THREADS=4`:

```
Starting 2 tests across 1 binary (1482 tests skipped)
PASS [  12.837s] arrow_linear_cofit_second_pass_is_a_noop_2023
[#2023 5a] block_ev=0.999948 arrow_ev=0.999983
PASS [ 199.911s] arrow_routed_linear_tier_matches_or_beats_block_reconstruction_2023
Summary [ 212.754s] 2 tests run: 2 passed, 1482 skipped
```

`arrow_routed_linear_tier_matches_or_beats_block_reconstruction_2023` **passes with the fix applied.** The regression e55ed1d33 reverted for is not caused by the change it reverted.

## Why the single post-land run could not have shown otherwise

That fixture is `n = 120` rows, and its certificate is `matches_mutable_state` — an exact `==` over the fitted decoder, frame, gauge singular values, logits and coords. #2512 measured that `SaeManifoldTerm::run_joint_fit_arrow_schur` **does not reproduce itself above ~61 rows**: bit-identical terms with bit-identical arguments give different decoders (22–31 of 40 entries differ, up to 56 ULP), with the threshold bisected to n=49..61 exact / n≥62 divergent. So an idempotent-re-entry certificate stated as bit-recurrence is, at n=120, comparing two draws from a nondeterministic distribution. A single red run of it is not evidence about any code change; a single green run is not evidence either, which is why I am reporting the pass together with the call-graph argument rather than on its own. Repeat runs of both arms are in flight.

## Consequences for this issue

- Option **(b)** stays dead, but so does the framing that made **(c)/(c')** necessary. The block-CG warm start in `update.rs:2627` is real, and it is genuinely a tolerance-not-bits situation — but it is not what `cofit_arrow` is measuring, because `cofit_arrow` never runs that solver. "One bug holding up two contracts in opposite directions" is two independent facts.
- The **cofit_arrow idempotence red is an independent defect** whose mechanism is #2512's nondeterminism, not a warm-started CG. A bit-idempotence predicate over a fitted quantity is unsatisfiable at that size until #2512 is fixed. That belongs with #2512 / the fit-acceptance work, not here.
- The fall-through is therefore unblocked and lands as 587fa42c7 stood. Verification of the red it fixes (`large_k_sparse_fit_stays_fixed_width_and_never_materializes_dense_n_by_k`, an integration test that a `--lib` filter silently skips) plus the `build.rs` audits are running now; I will land on that receipt.

**Comment by SauersML** [comment]
Current-main guarded production proof is now running on MSI as job `14168694` (CPU-only; one owned job, 12-minute ceiling).

Before submission I rechecked remote `main` at `f9db30315499098b47cca6189352d658fe669b0b`: both target preimages are still exact (`update.rs` blob `f672676050d84ad641c213be14791b7442c3f3da`; integration test blob `c764637a85e6aac3cf1841dcd16c6d70a6482751`). The staged patch/harness SHA-256 values also match the audited package exactly: `8c1ead4158dc8e...`, `87486dcb6f84d190...`, `bbe06925c4a816fa...`.

The harness takes the canonical nonblocking MSI lock, unsets the incompatible sccache layer, overlays exactly two files, rejects any extra changed path, and runs four exact optimized gates: dense-shortcut parity, matrix-free-CG parity, typed cap refusal, and the large-K non-vacuity route (dense decline > 0, CG work > 0, zero unresolved columns, certified residual). I will post the measured receipt and exact landed `main` commit only if all four pass and both target blobs remain unchanged at the end.

**Comment by SauersML** [comment]
Harness correction, with the failed run accounted for rather than hidden: MSI `14168694` stopped after 2m01s (peak RSS 6.80 GB). It guarded/overlaid the exact two files and compiled the candidate successfully—13 crates in 110s—but its first command named the stale module path `sparse_dict::update::tests::…`. The live module is `sparse_dict::update::exact_solve_tests`, so Rust honestly reported `running 0 tests`; the harness's cardinality gate rejected that immediately. No candidate test failed, and the remaining selectors did not run.

I changed only those three library selector prefixes. Candidate sources, both patches, test bodies, build configuration, and receipt criteria are unchanged. Corrected harness SHA-256 is `6a32c3085fb3811784f70c77600c21a4d6327fa903beb51b8008b9074ec7b919`; the staged copy matches exactly. Remote `main` remains `73cf09f0f0c8b7360c50497b4bbfec2cb6bab53c`, with both target preimages unchanged. The sole replacement proof is MSI job `14168791`.

**Comment by SauersML** [comment]
Resolved on direct `main` by `8e740e8136a8783f362c1f9e619c555fe22813e9`.

The root defect was false success: declined small-component Cholesky returned `Ok(())` without writing a solution. The landed implementation records the dense decline and continues through the existing matrix-free block-CG authority for the same normal equations. Only CG's typed stop and realized residual certify success; cap/breakdown retains the prior column and returns typed non-convergence. There is no ridge bump, diagonal substitute, fallback answer, or relaxation of the unrelated `cofit_arrow` recurrence contract.

MSI job `14168791` completed `0:0` in 3m41s (peak RSS 5.28 GB) on the exact landed two-file overlay. Four exact optimized gates passed:

1. dense-block parity — 1/1;
2. matrix-free CG vs independent dense solve at the charge floor — 1/1;
3. cap reached is typed non-convergence and never a substitute — 1/1;
4. large-K sparse integration: nonzero dense declines, positive CG work, zero unresolved columns, certified residual, no dense `N×K` routing state — 1/1.

The verifier held both target preimages across moving `main` (`a5445d8… → 2b1937c…`), and after the atomic ref update live `main` exposes exactly the proven postimage blobs: `update.rs` `8eee2eb484e8efcdf3bf3701ed956e1ae9a3f0a4`; integration test `80604707cb1b2110645b667fb16426d49b72a2c6`.

For completeness, `14168694` was a zero-test harness-selector rejection, not candidate evidence; it stopped immediately after successful compilation and was corrected transparently in the preceding comment.
